docs/cplusplus-style-guide.md

```diff
@@ -0,0 +1,585 @@
+# C++ code style guidelines
+
+We use C++14, but currently consider C++17 too immature for usage.
+
+This documents contains guidelines regarding:
+
+* File structure
+* Comments and documentation
+* Tests
+* Naming
+* Formatting
+* Typing
+* Namespaces
+* Structs
+* Classes
+* Error handling
+* Memory handling
+* Resource handling
+* Other
+
+
+## File structure
+
+* Header files: `.hpp`
+* Source files: `.cpp`
+* Use `#pragma once` in headers.
+* Split class definition across a header and source file by default.
+* Include header definitions in the same file.
+
+* Order of includes
+	1. Related header.
+	2. `C` library.
+	3. `C++` library.
+	4. Other libraries.
+	5. Fetch library.
+
+* Classes in this order
+	1. `public`.
+	2. `protected`.
+	3. `private`.
+
+
+## Comments and documentation
+
+* Provide API documentation with your code.
+* Include a comment summary at the top of all files.
+
+``` c++
+/* Approximation of the exponential function.
+   * @x is the argument that is used in the exponent 2^(x / M_LN2).
+   *
+   * We exploit the IEEE-754 standard to efficiently create a linear
+   * interpolation between points of the form 2^n where n is an integer.
+   *
+   * You can find the full details in
+   * [1] "A fast, compact approximation of the exponential function",
+   * Nicol N. Schraudolph, 1999 Neural computation 11, 853-862.
+   *
+   * @return the exponential of x.
+   */
+double Exp(double const &x) 
+{
+    // We use the constants explained in [1] and additional make e the
```

Comment:
Also we are using 2 space indents and that is consistent in this and other examples